### PR TITLE
Read MIDR_EL1 system register on aarch64

### PR DESCRIPTION
**Pull request by zorrorffm**
MIDR_EL1 system register exposes microarchitecture information so that people can make micro-arch related optimization such as exposing as much instruction level parallelism as possible.

MIDR_EL1 register can be read only if HWCAP_CPUID feature is supported.

We are going to optimize some algorithms (chacha20, AES-GCM etc.) based on this change which can get performance uplifted significantly. 

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by zorrorffm** [comment]
Does anyone take a look at this CL as well as the related one #12397? Thanks

**Comment by paulidale** [comment]
Ping for review.

File: crypto/armcap.c
Comment (ardbiesheuvel) [inline crypto/armcap.c:211]: I think we can drop this hunk: this code is not expected to execute at an exception level where MIDR_EL1 is accessible, and so the only way we can read it is through the CPUID emulation, which is a Linux specific feature, and which is advertised via the CPUID hwcap.

Comment (paulidale) [reply crypto/armcap.c:211]: I did wonder if it was necessary but figured it wasn't harmful.

Comment (ardbiesheuvel) [reply crypto/armcap.c:211]: Actually, it could be harmful, and it would be better to remove this code entirely at some point.

The problem is that the OS may decide to hide certain capabilities from user space, either because of silicon errata if there is a problem with the feature, or in the case of heterogeneous SMP (big.LITTLE), where a task may migrate from a CPU that has support for certain instructions to one that does not.

In this particular case, I think it is harmless, since the only possible way to access MIDR_EL1 is via the emulation, which will always produce the value the OS thinks is the correct one. But in general, we should abide by the hwcap features only, and avoid SIGILL trapping for feature discovery.

Comment (kroeckx) [reply crypto/armcap.c:211]: If the OS supports reading hardware capabilities, that is the only information we should use. But it's not available on all platforms, so we need fallback code.

The OS might not always provide enough information for us to make a good decision about which implementation to use. While a CPU might support 2 implementations we have, newer generation of CPUs might preform better using the other that worked best for the older generation. There can also be differences depending on the range of CPU: mobile vs. server. And as far as I know, we don't really either have code nor benchmark results for it to pick the best implementation.

There is 1 PR making use of this PR, which has a performance again of about 3%. I think we've previously considered 10% performance increase the minimum we should see to consider such changes.

Comment (ardbiesheuvel) [reply crypto/armcap.c]: In a sense, those are even worse: if sha256 is supported by the big CPUs, but not by the little CPUs, and the program happens to be started on big one, you might experience SIGILL crashes if the program gets migrated to a little CPU by the OS scheduler.

**Review by paulidale — APPROVED** [review]
Still approved.  Needs second review.

**Comment by ardbiesheuvel** [comment]
Not sure if my review counts or not, but I am fine with the patch as is. *However*, I am still not convinced we need this patch at all, given that the patch that relies on it is still under discussion, and is unlikely to be merged in its current form given the limited performance gain. So please don't merge this yet.

**Comment by zorrorffm** [comment]
> 
> 
> Not sure if my review counts or not, but I am fine with the patch as is. _However_, I am still not convinced we need this patch at all, given that the patch that relies on it is still under discussion, and is unlikely to be merged in its current form given the limited performance gain. So please don't merge this yet.

I agree that we can't merge a PR with limited performance uplift.  What I want to point out is that the original optimization on chacha20  is dedicated to ThunderX2, the author commented that his implementation is not friendly to most platforms but ThunderX2 because he thought ThunderX2 is server oriented, other platforms are client oriented. I believed that the author compromised because there wasn't a approach to detect the platform at run time. So actually this PR provide the possibility making chacha20 running at best performance on various platforms. Moreover we can see the performance data from his implementation, A53 gets 9% performance regression, this is to say, based on this PR, #12397 can revert the performance on A53. 

Another point I would like to mention is that this PR will benefit to  micro architecture  related optimization which maybe bring significant performance gain. Actually original chacha20 optimization on ThunderX2 is such an example, this optimization dedicated to ThunderX2 makes the performance 37% uplift.   Also for further ARM micro architecture like Zeus we will have more opportunities to optimize crypto algorithms based on this PR.  So I thinks this PR is valuable.

**Comment by zorrorffm** [comment]
Just ping...
This PR provides the opportunity to balance the performance on all platform.  Based on this PR,  #12397 makes all the performance on different platforms(A53, A55, A57, A72) uplifted while no performance regression on ThunderX2 occurs.

Thanks.

**Comment by paulidale** [comment]
ping @openssl/committers for review?

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. Thank you for the contribution.

**Pull request closed without merging**
